keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -563,6 +584,61 @@ class ApplicationService(
     )
   }
 
+  /**
+   * Adds details about any stateful constraints in the given environment to the [ArtifactSummaryInEnvironment].
+   * Also, adds details about any stateless constraints that have their status saved into the database
+   * (this happens when we approve an artifact version).
+   *
+   * For each stateful constraint type, if it's not yet been evaluated, creates a synthetic constraint summary object
+   * with a [ConstraintStatus.NOT_EVALUATED] status.
+   *
+   * Also, adds details about any stateless constraints that haven't already been populated
+   * in the given environment to the [ArtifactSummaryInEnvironment].
+   */
+  private fun ArtifactSummaryInEnvironment.addConstraintSummaries(
+    deliveryConfig: DeliveryConfig,
+     environment: Environment,
+     version: String,
+     artifact: DeliveryArtifact
+  ): ArtifactSummaryInEnvironment {
+    val persistedStates = repository.constraintStateFor(deliveryConfig.name, environment.name, version)
+    val notEvaluatedPersistedConstraints = environment.constraints.filter { constraint ->
+      constraint is StatefulConstraint && persistedStates.none { it.type == constraint.type }
+    }.map { constraint ->
+      StatefulConstraintSummary(
+        type = constraint.type,
+        status = NOT_EVALUATED
+      )
+    }
+
+    val statelessConstraintsStates: List<StatefulConstraintSummary> = environment.constraints.filter { constraint ->
+      // some/all stateless constraints might already have summary info, so filter those out.
+      constraint !is StatefulConstraint && persistedStates.none { it.type == constraint.type }
+    }.mapNotNull { constraint ->
+      statelessEvaluators.find { evaluator ->
+        evaluator.supportedType.name == constraint.type
+      }?.let { evaluator ->
+        val passes = evaluator.canPromote(artifact, version = version, deliveryConfig = deliveryConfig, targetEnvironment = environment)
+        StatefulConstraintSummary(
```

Comment:
Am I correct in understanding that we only need to build these summary objects here for backwards compatibility with stateless constraints that have already passed but won't have state stored in the DB before this PR?

Reply:
Yep!